src/client/Client.cc

```diff
@@ -9232,7 +9273,8 @@ int64_t Client::_write(Fh *f, int64_t offset, uint64_t size, const char *buf,
     }
   }
 
-  if (cct->_conf->client_oc && (have & CEPH_CAP_FILE_BUFFER)) {
+  if (cct->_conf->client_oc &&
+      (have & (CEPH_CAP_FILE_BUFFER | CEPH_CAP_FILE_LAZYIO))) {
```

Comment:
I don't like that `Fl` means that all writes/reads become synchronous. That is unnecessary and will hurt performance (which this feature is all about!). The need for two clients doing lazy I/O to have cache coherency will be the exception, not the norm. Let's change it so that the MDS can hand out `Fcbl` in this PR and, if it's easy, add another subtree xattr to have the MDS to **not** give out `Fcb` with `Fl`, if the user needs that.

Reply:
what's benefits of doing this. In single client case, mds issues Fcb to client even it wants Fl.

Reply:
> what's benefits of doing this.

So I wasn't thinking right. Forget what I said "I don't like that Fl means that all writes/reads become synchronous".

> In single client case, mds issues Fcb to client even it wants Fl.

I'm wondering if this could be changed so that `Fl` simply means that other clients may be doing reads/writes with no cache coherency guarantees. The MDS automatically hands out `Fb` with the `Fwl` cap... because the client **can** buffer writes.

In other words, for the client is there a real difference in the code between `Fwb` and `Fwl`. What if the MDS always gives you `Fb` with `Fl`? I feel this could be simpler and the `Fl` cap could just be checked by some `ceph_is_cache_coherent(int fd)` function.

/late night thinking. Please let me know if that doesn't make sense.

Reply:
Fb and Fl have difference.  Fb means client can do buffered write, Fl means client can do buffered write for files opened in lazy IO mode.

Issue Fb with Fl may make client code a little simpler. But it will mds code more complex. (move adjust_caps_used_for_lazyio logical to mds)

Reply:
> Fb and Fl have difference. Fb means client can do buffered write, Fl means client can do buffered write for files opened in lazy IO mode.

Logically only the MDS cares about LAZYIO as it permits handing out `Fbc` to more than one client in the presence of multiple writers. The client does not, itself, need to use LAZYIO cap at all except to ask the MDS to allow LAZYIO. However, keeping LAZYIO as a cap still makes sense to simplify code in the MDS; however, it does not modify the client's behavior.

> Issue Fb with Fl may make client code a little simpler. But it will make mds code more complex. (move adjust_caps_used_for_lazyio logical to mds)

Is it really that complex though? We're just going to (a) add `Fb` if `Flw` and (b) add `Fc` if `Flr`.

We do not need to worry about backwards-compatibility so I would prefer we do this in a way that makes sense.